language/move-vm/runtime/src/data_cache.rs

```diff
@@ -44,21 +53,30 @@ impl ConfigStorage for &dyn RemoteCache {
 /// The Move VM takes a `DataStore` in input and this is the default and correct implementation
 /// for a data store related to a transaction. Clients should create an instance of this type
 /// and pass it to the Move VM.
-pub struct TransactionDataCache<'txn> {
-    data_map: BTreeMap<AccessPath, Option<(FatStructType, GlobalValue)>>,
-    module_map: BTreeMap<ModuleId, Vec<u8>>,
-    event_data: Vec<ContractEvent>,
-    data_cache: &'txn dyn RemoteCache,
+pub(crate) struct TransactionDataCache<'r, 'l, C> {
+    data_cache: &'r C,
+    loader: &'l Loader,
+    account_map: BTreeMap<AccountAddress, AccountDataCache>,
+    event_data: Vec<(Vec<u8>, u64, Type, Value)>,
 }
 
-impl<'txn> TransactionDataCache<'txn> {
+pub struct TransactionEffects {
+    pub resources: Vec<(
+        AccountAddress,
+        Vec<(TypeTag, Option<(MoveTypeLayout, Value)>)>,
+    )>,
+    pub modules: Vec<(ModuleId, Vec<u8>)>,
+    pub events: Vec<(Vec<u8>, u64, TypeTag, MoveTypeLayout, Value)>,
+}
+
+impl<'r, 'l, C: RemoteCache> TransactionDataCache<'r, 'l, C> {
     /// Create a `TransactionDataCache` with a `RemoteCache` that provides access to data
     /// not updated in the transaction.
-    pub fn new(data_cache: &'txn dyn RemoteCache) -> Self {
+    pub(crate) fn new(data_cache: &'r C, loader: &'l Loader) -> Self {
```

Comment:
this makes me think that a cache separate from the loader would be better

Reply:
That is an interesting idea. I'd actually think about this within a bigger context: how do we view the `Type`s stored in the `TransactionDataCache`? There are two possible approaches I can see here:
1) `TransactionDataCache` must know the meanings of a `Type`, which means it must contain a reference to the loader. 
2) `TransactionDataCache` treats `Type`s as black boxes and only use them as keys. My guess is that this will make the `TransactionDataCache` more light-weighted as we'll move some logic to the callee positions (`Session` for example). 

Anyways one advantage we have right now is that `TransactionDataCache` is a private to the crate so it should be much easier for us to change it in the future.

Reply:
I think 2) is what we should do. Basically that data cache treats Types as keys and if the layout was not loaded yet the loader is accessed to get the info to create a layout.
But anyway that is obviously for another